src/java.base/share/classes/java/io/Reader.java

```diff
@@ -203,9 +203,7 @@ public int read(char[] cbuf, int off, int len) throws IOException {
             public String readAllAsString() throws IOException {
                 ensureOpen();
                 int len = cs.length();
-                char[] cbuf = new char[len - next];
-                cs.getChars(next, len, cbuf, 0);
-                String result = new String(cbuf);
+                String result = cs.subSequence(next, len).toString();
```

Comment:
Are you sure that `subSequence` is more efficient that `getChars`? Otherwise I do not see why replacing it.

Reply:
I think this may be a good way to reduce redundant allocation if the CharSequence's subsequence is a simple view (frequently the case for user ones) or a String (for StringBuilder, String)

Reply:
I have no idea whether `subSequence` is more performant than `getChars` but it seems cleaner. In any case, at this point we are concerned primarily with the APIs getting into JDK 25. We can improve performance after that.

Reply:
Actually I am working on exact these optimizations right now and do not like the idea that I have to file another JBS and PR just for a single code line. As `subSequence` typically *does* a copy (to not hold the original full text in-memory for the lifetime of a possibly single-character sub sequence), unless otherwise proven, `getChars` will always be faster. Having said that, I need to say, that I do not second what Chen said. In reality I have never seen "simple views", in particular not "frequently".

Reply:
BTW, paraphrasing Stuart here: "We want to reduce self-calls". `subSequence` is a self-call, *as we do not know* how it behaves in the actual implementation. For `getChars` we can be sure that it has no overridable side effects.

Reply:
Both `subSequence` and `getChars` are implemented by the CharSequence.
They have the same level of trustworthiness about their implementation.

Reply:
I meant that the implementation of `subSequence` *might or might not* perform copy (so we need to tell the javaDoc reader), while `getChar` *clearly* performs a copy as part of it design. So `subSequence` can be surprising as this discussion proofs, while `getChars` will not. That was Stuart's point with `read`, and that was your point with `CharSequence`.

Reply:
@mkarg Please don't invoke my name to try to buttress your arguments. The calls the Reader.of() instance makes on its backing CharSequence is a different kind of issue from what promises or guarantees the concrete methods of the Reader class makes to its subclasses.

Reply:
@stuart-marks Sorry, didn't want to pull you in here, that's why I said, I am just *paraphrasing*.

That is correct, it is a different case, but the *final effect* for the caller is the same: If he needs to now whether a copy is made *or not*, he needs to have JavaDocs telling him *how* the new method works inside (whether it calls `subSequence` or whether it calls `getChars`). Otherwise he might make false assumptions about the result.

Reply:
This API is concerned with the "what" and not so much the "how."

Reply:
The purpose of Reader and subclasses is to normalize characters in various forms (char arrays, byte streams, char buffers, etc) into a coherent stream of characters and be able to read it incrementally.
If the caller is going to use the entire CharSequence, as in the case of `Reader.of(cs)`, there is no reason to have a reader at all. Both String and CharSequence have more useful APIs than Reader for parsing.
The motivation for reading all characters and reading lines came from use cases around input from sub-processes where the input is decoded from a byte stream, but that's not `Reader.of(cs).`

Reply:
In many situations the programmers has no choice whether to use a reader or not, as the code he works is must fulfilf a given specification (like JAX-RS), so he is getting a `Reader` *whether needed or not* -- it is not under *his* control. But as he wants to get *all* content, he likes to use this new method. Pretty valid and frequent use case in the enterprise world.